### PR TITLE
catalog-backend: initial implementation of relations storage

### LINKED ISSUE #2958: catalog-backend: add migration and db layer for relations
Part of #1964 

Implement as described in #2617

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Fixes #2958 

This adds the initial implementation of entity relations storage. It does not implement querying and needs a few more tests before it's ready for review.

Once this is in we can continue #2617 to get the it set up e2e

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/catalog-backend/src/database/CommonDatabase.ts [hunk 1]
```diff
@@ -330,6 +335,33 @@ export class CommonDatabase implements Database {
...
+  ): Promise<void> {
+    const tx = txOpaque as Knex.Transaction<any, any>;
+
+    // remove all relations that exist for the originating entity id.
+    await tx<DbEntitiesRelationsRow>('entities_relations')
+      .where({ originating_entity_id: originatingEntityId })
+      .del();
+
+    const serializeName = (e: EntityName) =>
+      `${e.kind}:${e.namespace}/${e.name}`.toLowerCase();
```
Comment (freben) [inline plugins/catalog-backend/src/database/CommonDatabase.ts:351]: This lowercasing thing, I've been thinking if that should ALWAYS happen (in the catalog-model functions) or not (if it should be implementation dependent like here)

File: plugins/catalog-backend/migrations/20201019130742_add_relations_table.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    table
+      .string('type')
+      .notNullable()
+      .comment('The type of the relation between the entities');
+    table
+      .string('target_full_name')
+      .notNullable()
+      .comment('The full name of the target entity of the relation');
+
+    table.primary(['source_full_name', 'type', 'target_full_name']);
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-backend/migrations/20201019130742_add_relations_table.js:45]: I mean, we already delete all and replace

File: plugins/catalog-backend/src/database/CommonDatabase.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/catalog-backend/src/database/CommonDatabase.ts:351]: Yup, definitely. Beginning to feel I'd rather be passing around the serialized entity name too

Comment (freben) [reply plugins/catalog-backend/src/database/CommonDatabase.ts:351]: Hmm. Maybe EntityName should actually be a class with `matches(Entity)` (case insensitively), `toLowercaseString()`, and `from(String)`

File: packages/catalog-model/src/entity/Entity.ts [hunk 3]
```diff
@@ -120,3 +126,37 @@ export type EntityMeta = JsonObject & {
...
+  type: string;
+
+  /**
+   * The target entity of this relation.
+   */
+  target: EntityName;
+};
+
+/**
+ * Holds the relationship data for entities
```
Comment (freben) [inline packages/catalog-model/src/entity/Entity.ts:146]: relation

File: packages/catalog-model/src/entity/Entity.ts [hunk 4]
```diff
@@ -120,3 +126,37 @@ export type EntityMeta = JsonObject & {
...
+ */
+export type EntityRelationSpec = {
+  /**
+   * The source entity of this relation.
+   */
+  source: EntityName;
+  /**
```
Comment (freben) [inline packages/catalog-model/src/entity/Entity.ts:153]: newline :)

File: packages/catalog-model/src/entity/Entity.ts (hunk 3 above)
Comment (freben) [reply packages/catalog-model/src/entity/Entity.ts:146]: also dot at end

**Review by benjdlambert — APPROVED** [review]

File: plugins/catalog-backend/migrations/20201019130742_add_relations_table.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+exports.up = async function up(knex) {
+  await knex.schema.createTable('entities_relations', table => {
+    table.comment('All relations between entities in the catalog');
+    table
+      .uuid('originating_entity_id')
+      .references('id')
+      .inTable('entities')
+      .onDelete('CASCADE')
+      .notNullable()
+      .comment('The originating entity of the relation');
```
Comment (freben) [inline plugins/catalog-backend/migrations/20201019130742_add_relations_table.js:31]: maybe the "originating" word is a bit fuzzy ... i guess the description text could be a bit more down to earth like "The entity that caused this relation to be generated"

File: plugins/catalog-backend/migrations/20201019130742_add_relations_table.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      .notNullable()
+      .comment('The type of the relation between the entities');
+    table
+      .string('target_full_name')
+      .notNullable()
+      .comment('The full name of the target entity of the relation');
+
+    table.primary(['source_full_name', 'type', 'target_full_name']);
+  });
+};
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-backend/migrations/20201019130742_add_relations_table.js:47]: I still think we need the index!

[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-backend/migrations/20201019130742_add_relations_table.js:47]: I have changed my mind!

**Pull request merged**